website/blog/2023-03-24-databend-weekly-86.md

```diff
@@ -0,0 +1,150 @@
+
+- [Doc | AI Functions - AI_TO_SQL](https://databend.rs/doc/sql-functions/ai-functions/ai-to-sql)
+- [Blog | Databend Understands You Better with OpenAI](https://databend.rs/blog/ai2sql)
+
+## Code Corner
+
+Discover some fascinating code snippets or projects that showcase our work or learning journey.
+
+### Vector Similarity Calculation in Databend
+
+Databend adds a new function called `cosine_distance`. This function accepts two input vectors, `from` and `to`, which are represented as slices of f32 values.
+
+```sql
+select cosine_distance([3.0, 45.0, 7.0, 2.0, 5.0, 20.0, 13.0, 12.0], [2.0, 54.0, 13.0, 15.0, 22.0, 34.0, 50.0, 1.0]) as sim
+----
+0.8735807
+```
+
+The Rust implementation efficiently performs calculations by utilizing the `ArrayView` type from the [ndarray](https://crates.io/crates/ndarray) crate.
+
+```rust
+pub fn cosine_distance(from: &[f32], to: &[f32]) -> Result<f32> {
+    if from.len() != to.len() {
+        return Err(ErrorCode::InvalidArgument(format!(
+            "Vector length not equal: {:} != {:}",
+            from.len(),
+            to.len(),
+        )));
+    }
+
+    let a = ArrayView::from(from);
+    let b = ArrayView::from(to);
+    let aa_sum = (&a * &a).sum();
+    let bb_sum = (&b * &b).sum();
+
+    Ok((&a * &b).sum() / ((aa_sum).sqrt() * (bb_sum).sqrt()))
+}
+```
+
+Do you remember how to register scalar functions in Databend? You can check [Doc | How to Write a Scalar Function](https://databend.rs/doc/contributing/how-to-write-scalar-functions) and [PR | #10737](https://github.com/datafuselabs/databend/pull/10737) to verify your answer.
+
+## Highlights
+
+Here are some noteworthy items recorded here, perhaps you can find something that interests you.
+
+- Learn how to monitor Databend using Prometheus and Grafana by reading our new document. *[Doc | Monitor - Prometheus & Grafana](https://databend.rs/doc/monitor/prometheus-and-granafa)*
+- If you want to connect Metabase with Databend and create dashboards, we have developed the [Metabase Databend Driver](https://github.com/databendcloud/metabase-databend-driver/releases/latest) and provided simple steps. *[Doc | Integrations - Metabase](https://databend.rs/doc/integrations/gui-tool/metabase)*
+- Databend now supports `PIVOT`, `UNPIVOT`, `GROUP BY CUBE` and `GROUP BY ROLLUP` query syntax. For more information, please see PR [#10676](https://github.com/datafuselabs/databend/pull/10676) and [#10601](https://github.com/datafuselabs/databend/pull/10601).
+
+## What's Up Next
+
+We're always open to cutting-edge technologies and innovative ideas. You're more than welcome to join the community and bring them to Databend.
+
+### Enable `-Zgitoxide` to Speed up Git Dependencies Download
+
+To accelerate the download of our git dependencies, we can activate `-Zgitoxide`. This option is significantly faster than using Git alone.
```

Comment:
Enabling `-Zgitoxide` can speed up the download of our Git dependencies significantly, which is much faster than using Git only.